File: stepik-java-api/src/main/java/org/stepik/api/queries/auth/AbstractAuthorizationPostQuery.java
package org.stepik.api.queries.auth;

import org.jetbrains.annotations.NotNull;
import org.stepik.api.actions.StepikAbstractAction;
import org.stepik.api.objects.auth.TokenInfo;
import org.stepik.api.queries.StepikAbstractPostQuery;
import org.stepik.api.urls.Urls;

/**
 * @author meanmail
 */
public abstract class AbstractAuthorizationPostQuery extends StepikAbstractPostQuery<TokenInfo> {
    public AbstractAuthorizationPostQuery(@NotNull StepikAbstractAction stepikAction) {
        super(stepikAction, TokenInfo.class);
    }

    @NotNull
    @Override
    protected String getUrl() {
        return Urls.OAUTH_URL;
    }

    @NotNull
    @Override
    public TokenInfo execute() {
        TokenInfo tokenInfo = super.execute();

        StepikAbstractAction action = getStepikAction();
        action.getStepikApiClient().setTokenInfo(tokenInfo);

        return tokenInfo;
    }

    @NotNull
    @Override
    protected String getContentType() {
        return "application/x-www-form-urlencoded";
    }
}

File: stepik-java-api/src/main/kotlin/org/stepik/api/queries/auth/AbstractAuthorizationPostQuery.kt
package org.stepik.api.queries.auth

import org.stepik.api.actions.StepikAbstractAction
import org.stepik.api.objects.auth.TokenInfo
import org.stepik.api.queries.StepikAbstractPostQuery

abstract class AbstractAuthorizationPostQuery(stepikAction: StepikAbstractAction) :
        StepikAbstractPostQuery<TokenInfo>(stepikAction, TokenInfo::class.java) {
    
    override val url = "${stepikAction.stepikApiClient.host}/oauth2/token/"
    
    override fun execute(): TokenInfo {
        val tokenInfo = super.execute()
        
        val action = stepikAction
        action.stepikApiClient.tokenInfo = tokenInfo
        
        return tokenInfo
    }
    
    override val contentType = "application/x-www-form-urlencoded"
    
}
